test/hotspot/jtreg/runtime/locking/TestRecursiveMonitorChurn.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/*
+ * @test
+ * @summary Tests that recursive locking doesn't cause excessive native memory usage
+ * @library /test/lib
+ * @run main TestRecursiveMonitorChurn
+ */
+public class TestRecursiveMonitorChurn {
+    static class Monitor {
+        volatile int i, j;
+        synchronized void doSomething() {
+            i++;
+            doSomethingElse();
+        }
+        synchronized void doSomethingElse() {
+            j++;
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
+        if (args.length == 1 && args[0].equals("test")) {
+            // The actual test, in a forked JVM.
+            for (int i = 0; i < 100000; i++) {
+                Monitor m = new Monitor();
+                m.doSomething();
+            }
+        } else {
+            ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(
```

Comment:
Are there any reasons to use createLimitedTestJavaProcessBuilder here?
If test should work with C1/C2 and different locking modes then it would be better just to use createTestJavaProcessBuilder and don't use 'TieredStopAtLevel'. So the test could be executed with C1/C2, different GC, and any other VM flags.
They should be combined with test vm options.

Reply:
createLimitedTestJavaProcessBuilder is a copy+paste mistake ;-) Running with TieredStopAtLevel is intentional, I think C2 would eliminate the locks and might lead to false positives. OTOH, the test would run long enough in interpreter and C1 to still exceed the monitor threshold, so should be ok.